File: ROS/day1/hw2/hw2_package/publisher.cpp
#include <chrono>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/int32.hpp>
#include <std_msgs/msg/string.hpp>
#include <std_msgs/msg/float32.hpp>

using namespace std::chrono_literals;

class MultiPublisher : public rclcpp::Node {
public:
  MultiPublisher() : Node("multi_publisher"), count_(0) {
    pub_int_ = create_publisher<std_msgs::msg::Int32>("int_topic", 10);
    pub_str_ = create_publisher<std_msgs::msg::String>("string_topic", 10);
    pub_flt_ = create_publisher<std_msgs::msg::Float32>("float_topic", 10);

    timer_ = create_wall_timer(1s, [this] { publish_msgs(); });
  }

private:
  void publish_msgs() {
    auto i = std_msgs::msg::Int32();
    i.data = count_;

    auto s = std_msgs::msg::String();
    s.data = "Hello " + std::to_string(count_);

    auto f = std_msgs::msg::Float32();
    f.data = 3.14f * count_;

    pub_int_->publish(i);
    pub_str_->publish(s);
    pub_flt_->publish(f);

    RCLCPP_INFO(get_logger(), "Published -> int=%d, str=%s, float=%.2f",
                i.data, s.data.c_str(), f.data);

    count_++;
  }

  rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr pub_int_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_str_;
  rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr pub_flt_;
  rclcpp::TimerBase::SharedPtr timer_;
  int count_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MultiPublisher>());
  rclcpp::shutdown();
  return 0;
}

File: ROS/day1/hw2/hw2_package/publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int32, String, Float32

class MultiPublisher(Node):
    def __init__(self):
        super().__init__('multi_publisher')
        self.pub_int = self.create_publisher(Int32, 'int_topic', 10)
        self.pub_str = self.create_publisher(String, 'string_topic', 10)
        self.pub_flt = self.create_publisher(Float32, 'float_topic', 10)
        self.count = 0
        self.timer = self.create_timer(1.0, self.timer_callback)

    def timer_callback(self):
        msg_int = Int32()
        msg_int.data = self.count

        msg_str = String()
        msg_str.data = f'Hello {self.count}'

        msg_flt = Float32()
        msg_flt.data = 3.14 * self.count

        self.pub_int.publish(msg_int)
        self.pub_str.publish(msg_str)
        self.pub_flt.publish(msg_flt)

        self.get_logger().info(
            f'Published -> int={msg_int.data}, str={msg_str.data}, float={msg_flt.data:.2f}'
        )

        self.count += 1


def main(args=None):
    rclpy.init(args=args)
    node = MultiPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
